### PR TITLE
[NET-4144] add gateway pdb

### PR DESCRIPTION
**Pull request by kkavish**
Changes proposed in this PR:
- Add PDB for all gateways
- 

How I've tested this PR:
manually

How I expect reviewers to test this PR:
- Deploy consul k8s cluster using kind with gateways, check kubectl get poddisruptionbudget -n <namespace> for PDBs
- Unit tests

Checklist:
- [x] Tests added
- [x] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+## Unplanned Release
+
+BUG FIXES:
```
Comment (Ganeshrockz) [inline CHANGELOG.md:3]: I don't think you should update this file. Adding `.changelog/*.txt` should be suffice.

File: .changelog/2985.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
```
Comment (Ganeshrockz) [inline .changelog/2985.txt:1]: Is this a bug? Should we call it as an enhancement?

File: charts/consul/values.yaml [hunk 3]
```diff
@@ -66,7 +66,7 @@ global:
   # image: "hashicorp/consul-enterprise:1.10.0-ent"
   # ```
   # @default: hashicorp/consul:<latest version>
-  image: docker.mirror.hashicorp.services/hashicorppreview/consul:1.17-dev
+  image: docker.mirror.hashicorp.services/hashicorppreview/consul:1.16.1
```
Comment (Ganeshrockz) [inline charts/consul/values.yaml:69]: I guess this should still be `1.17-dev`

File: charts/consul/values.yaml [hunk 4]
```diff
@@ -2272,7 +2272,7 @@ connectInject:
     # If true, then all traffic redirection setup uses the consul-cni plugin.
     # Requires connectInject.enabled to also be true.
     # @type: boolean
-    enabled: false
+    enabled: true
```
Comment (Ganeshrockz) [inline charts/consul/values.yaml:2275]: Any reason to enable CNI?

File: charts/consul/values.yaml [hunk 5]
```diff
@@ -2138,7 +2138,7 @@ connectInject:
   # [injection annotation](https://developer.hashicorp.com/consul/docs/k8s/connect#consul-hashicorp-com-connect-inject)
   # to opt-in to Connect injection. If this is true, pods can use the same annotation
   # to explicitly opt-out of injection.
-  default: false
+  default: true
```
Comment (Ganeshrockz) [inline charts/consul/values.yaml:2141]: Is this change required?

File: charts/consul/values.yaml [hunk 6]
```diff
@@ -3098,6 +3119,22 @@ terminatingGateways:
...
+    # Enables registering a PodDisruptionBudget for the terminating gateways
+    # If enabled, it only registers the budget so long as
+    # the terminating gateways is enabled. To disable, set to `false`.
+    enabled: true
+
+    # The maximum number of unavailable pods. By default, this will be
+    # automatically computed based on the `server.replicas` value to be `(n/2)-1`.
+    # If you need to set this to `0`, you will need to add a
+    # --set 'server.terminatingGateways.maxUnavailable=0'` flag to the helm chart installation
+    # command because of a limitation in the Helm templating language.
```
Comment (Ganeshrockz) [inline charts/consul/values.yaml:3134]: > command because of a limitation in the Helm templating language.

What limitation is this?

File: charts/consul/values.yaml [hunk 7]
```diff
@@ -2753,6 +2753,27 @@ meshGateway:
     # DNS entry to point to your mesh gateways.
     static: ""
 
+  # This configures the [`PodDisruptionBudget`](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
```
Comment (Ganeshrockz) [inline charts/consul/values.yaml:2756]: May be I am missing something, but I don't see a `disruptionBudget` stanza for `apiGateways` but it seems to be referenced in the helm template files. Is this intentional?

File: charts/consul/templates/crd-gateways-disruptionbudget.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
+{{- if .Values.connectInject.enabled }}
```
Comment (Ganeshrockz) [inline charts/consul/templates/crd-gateways-disruptionbudget.yaml:1]: What are CRD gateways?

**Review by Ganeshrockz — COMMENTED** [review]
Skimmed through the changes and have some basic questions. Will go through the PR again in detail once those queries get resolved.

File: charts/consul/templates/crd-gateways-disruptionbudget.yaml (hunk 8 above)
Comment (kkavish) [reply charts/consul/templates/crd-gateways-disruptionbudget.yaml:1]: custom gateways,  name: gateways.gateway.networking.k8s.io

File: charts/consul/values.yaml (hunk 7 above)
Comment (kkavish) [reply charts/consul/values.yaml:2756]: please check line #3276, this is for consul api gateway, connect inject uses CRD gateway as per my understanding, it has its own default pdb yaml.

File: charts/consul/values.yaml (hunk 6 above)
Comment (kkavish) [reply charts/consul/values.yaml]: ah, seems to be a copied comment not applicable here

File: CHANGELOG.md (hunk 1 above)
Comment (kkavish) [reply CHANGELOG.md]: noted.

File: .changelog/2985.txt (hunk 2 above)
Comment (kkavish) [reply .changelog/2985.txt:1]: reported as a bug, I too had this confusion.

File: charts/consul/values.yaml (hunk 3 above)
Comment (kkavish) [reply charts/consul/values.yaml]: yes, it was for local testing, reverting

File: charts/consul/values.yaml (hunk 4 above)
Comment (kkavish) [reply charts/consul/values.yaml]: nope, it was for local testing, reverting

File: charts/consul/values.yaml (hunk 5 above)
Comment (kkavish) [reply charts/consul/values.yaml]: nope, it was for local testing, reverting

File: charts/consul/values.yaml (hunk 7 above)
Comment (Ganeshrockz) [reply charts/consul/values.yaml:2756]: Ahh my bad 🤦 . The diff confused me. Thanks for pointing out

File: .changelog/2985.txt (hunk 2 above)
Comment (Ganeshrockz) [reply .changelog/2985.txt:1]: It could be an improvement

**Review by Ganeshrockz — COMMENTED** [review]
Looks good mostly. Deferring the approval to the `service-mesh`/`gateways` team who owns the codebase.

File: .changelog/2985.txt (hunk 2 above)
Comment (curtbushko) [reply .changelog/2985.txt:1]: I'd call it an 'improvement' also.

Comment (kkavish) [reply .changelog/2985.txt]: updated

**Comment by mr-miles** [comment]
Sorry to drop in out of nowhere!  But any chance to add a pdb for the ingress gateways too?  I know ingress gateways themselves are deprecated for replacement by api-gateways, but they are still supported.  Its confusing if the gateways are inconsistent in their config, unless they're going to be removed from the chart completely.

**Comment by kkavish** [comment]
> Sorry to drop in out of nowhere! But any chance to add a pdb for the ingress gateways too? I know ingress gateways themselves are deprecated for replacement by api-gateways, but they are still supported. Its confusing if the gateways are inconsistent in their config, unless they're going to be removed from the chart completely.

-  thanks for suggesting it, makes sense, we'd add it to ingress as well.

File: charts/consul/templates/crd-gateways-disruptionbudget.yaml
Comment (nathancoleman) [inline charts/consul/templates/crd-gateways-disruptionbudget.yaml:1]: Based on the resolved comment below, it seems like this is intended to target Consul API Gateway pods resulting from a user creating a `Gateway` resource like this:

```yaml
apiVersion: gateways.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: my-gateway
spec:
  gatewayClassName: nathan-gateway
```

This won't work in practice though and doesn't match how Gateways are structured.
Gateways specify a `GatewayClass`. For Consul, each `GatewayClass` has a `GatewaysClassConfig` (which we define -- [docs](https://developer.hashicorp.com/consul/docs/api-gateway/configuration/gatewayclassconfig)) tied to it that controls configuration specific to that `GatewayClass` ([docs](https://gateway-api.sigs.k8s.io/api-types/gatewayclass/)). You can have up to `n` GatewayClasses, each with their own GatewayClassConfig.

I'd expect this configuration to live on the `GatewayClassConfig` defined [here](https://github.com/hashicorp/consul-k8s/blob/012227d5f0ebf23f582e6f0301e73d161e1a3863/charts/consul/templates/crd-gatewayclassconfigs.yaml). The resulting GatewayClassConfig with support for `PodDisruptionBudget` might look like:

```yaml
apiVersion: consul.hashicorp.com/v1alpha1
kind: GatewayClassConfig
metadata:
  name: nathan-gateway
spec:
  podDisruptionBudget:
    maxUnavailable: 2
  ...
```

API Gateway resources -- such as Deployment, Service, ServiceAccount, etc. -- are then created in K8s dynamically when a user creates a `Gateway` resource. I'd expect the logic to create a `PodDisruptionBudget` to live in [the `gatekeeper` package](https://github.com/hashicorp/consul-k8s/tree/012227d5f0ebf23f582e6f0301e73d161e1a3863/control-plane/api-gateway/gatekeeper) in a new file called `poddisruptionbudget.go`.

File: charts/consul/values.yaml [hunk 9]
```diff
@@ -2753,6 +2753,25 @@ meshGateway:
...
+    # The maximum number of unavailable pods. By default, minAvailable will be set to 1
+    # If you need to set this to some value, you will need to add a
+    # --set 'meshGateway.disruptionBudget.maxUnavailable=0'` flag to the helm chart installation
+    # @type: integer
+    maxUnavailable: null
+
+    # The minimum number of available pods.
+    # Takes precedence over maxUnavailable if set.
+    # @type: integer
+    minAvailable: null
```
Comment (nathancoleman) [inline charts/consul/values.yaml:2773]: I don't see this value actually being consumed anywhere. I'd expect we either:
1. allow the user to set `minAvailable` for every gateway type, or
2. never allow the user to set `minAvailable` for any gateway type

**Review by nathancoleman — CHANGES_REQUESTED** [review]

File: control-plane/api-gateway/gatekeeper/poddisruptionbudget.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+func (g *Gatekeeper) upsertPodDisruptionBudget(ctx context.Context, gateway gwv1beta1.Gateway, gcc v1alpha1.GatewayClassConfig, config common.HelmConfig) error {
+	if gcc.Spec.PodDisruptionBudgetSpec == nil || !gcc.Spec.PodDisruptionBudgetSpec.Enabled {
+		return g.deletePodDisruptionBudget(ctx, types.NamespacedName{Namespace: gateway.Namespace, Name: gateway.Name})
+	}
+
+	pdb := g.podDisruptionBudget(ctx, gateway, gcc, config)
+	mutated := pdb.DeepCopy()
+	mutator := newPodDisruptionBudgetMutator(pdb, mutated, gateway, g.Client.Scheme())
+
+	result, err := controllerutil.CreateOrUpdate(ctx, g.Client, mutated, mutator)
```
Comment (nathancoleman) [inline control-plane/api-gateway/gatekeeper/poddisruptionbudget.go:27]: Let's follow the pattern found in `role.go` where we check for an existing `PodDisruptionBudget` and, if one exists, verify that it's owned by our controller.

In the future, I'd like to use that pattern for all resources that we create, but that is outside the scope of this PR.

**Review by nathancoleman — CHANGES_REQUESTED** [review]

File: control-plane/api-gateway/gatekeeper/poddisruptionbudget.go (hunk 10 above)
Comment (kkavish) [reply control-plane/api-gateway/gatekeeper/poddisruptionbudget.go:27]: my thought process was:

The configs remain same for resources that we create and the resources created through templates. i.e. they all have - enable, minAvailable and maxUnavailable thus making them homogenous.
For the resources we create, if the end-user/dev needs to disable PDB she/he could just go ahead and make enabled = false instead of deleting the whole stanza and adding it back again when they require it.
Shall I go ahead and remove the enabled flag anyways?

File: control-plane/api/v1alpha1/api_gateway_types.go [hunk 11]
```diff
@@ -65,6 +66,15 @@ type GatewayClassConfigSpec struct {
 
 	// The value to add to privileged ports ( ports < 1024) for gateway containers
 	MapPrivilegedContainerPorts int32 `json:"mapPrivilegedContainerPorts,omitempty"`
+
+	// The value to add pod disruption budgets, will prevent gateway pods from being evicted all at once.
+	PodDisruptionBudgetSpec *PodDisruptionBudgetSpec `json:"podDisruptionBudget,omitempty"`
+}
+
+type PodDisruptionBudgetSpec struct {
+	Enabled        bool                `json:"enabled,omitempty"`
```
[the remark this answers is not in the document]
Comment (kkavish) [reply control-plane/api/v1alpha1/api_gateway_types.go:75]: my thought process was:

1. The configs remain same for resources that we create and the resources created through templates. i.e. they all have - enable, minAvailable and maxUnavailable thus making them homogenous.
2. For the resources we create, if the end-user/dev needs to disable PDB she/he could just go ahead and make enabled = false instead of deleting the whole stanza and adding it back again when they require it.

Shall I go ahead and remove the enabled flag anyways?

**Comment by david-yu** [comment]
@nathancoleman I chatted with @kkavish he said this might need re-review? However I'm not sure if it's ready given that it looks like most of the tests are still failing?

**Comment by nathancoleman** [comment]
Re-reviewing is on my To Do list; however, it's lower in priority than work related to Consul 1.17.0 releasing next week. I confirmed with @kkavish that we're **not** trying to get this change into 1.17.0.

**Comment by david-yu** [comment]
Sure not a rush. Let's revisit after 1.17, I also asked @kkavish to brush up this PR to get the tests to pass as well.

**Comment by david-yu** [comment]
@kkavish Could you try re-basing to get the unit tests on control plane to pass?

**Comment by david-yu** [comment]
I'm guess the PR is still failing to pass any acceptance tests?

**Comment by kkavish** [comment]
> I'm guess the PR is still failing to pass any acceptance tests?

@david-yu sorry, I've been on a vacation, I'll try to get this fixed this week.
